### PR TITLE
IPv6 in IPv4 encapsulation

### PR DESCRIPTION
**Pull request by mchtech**
sit module can support both IPv4-in-IPv4 and IPv6-in-IPv4 tunnel.

### IPIP decap
```bash
ip link add name vip type sit local ${local4}
ip addr add ${vip4}/32 dev vip
ip addr add ${vip6}/128 dev vip
ip link set vip up
```

|  ipip(sit)                              | 4in4 | 4in6 | 6in4 | 6in6 |
| ------------------------------ | ---- |----- | ---- | ---- |
| 3.10.0-1160 (centos 7)         |  √   |  √   |  √   |  √   |
| 4.15.0-136  (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 4.18.0-24   (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 4.18.0-227  (centos 8)         |  √   |  √   |  √   |  √   |
| 5.4.96-200  (centos 7 AltArch) |  √   |  √   |  √   |  √   |
| 5.4.0-66    (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 5.11.1      (elrepo)           |  √   |  √   |  √   |  √   |

### GUE decap
```bash
modprobe fou
ip fou add port 6080 gue
ip link add name vip type sit encap gue encap-dport 6080 encap-sport auto local ${local4}
ip addr add ${vip4}/32 dev vip
ip addr add ${vip6}/128 dev vip
ip link set vip up
```

|   gue(sit)                             | 4in4 | 4in6 | 6in4 | 6in6 |
| ------------------------------ | ---- |----- | ---- | ---- |
| 3.10.0-1160 (centos 7)         |no fou|no fou|no fou|no fou|
| 4.15.0-136  (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 4.18.0-24   (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 4.18.0-227  (centos 8)         |no fou|no fou|no fou|no fou|
| 5.4.96-200  (centos 7 AltArch) |  √   |  √   |  √   |  √   |
| 5.4.0-66    (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 5.11.1      (elrepo)           |  √   |  √   |  √   |  √   |

## healthchecking

> we can only use direct healthchecking for IPv6-in-IPv4 encap, because it doesn't work well with healthchecking_ipip.

### ipip encap by healthchecking_ipip.o
| healthchecking_ipip.o         | ipip (4in4) | ipip (6in4) | sit (4in4)     | sit (6in4)   | ip6tnl (6in4)| ip6tnl (6in6) |
| ----------------------------- | ----------- | ----------- | -------------- | ------------ | ------------ | ------------- |
| 4.15.0-136 (ubuntu 18)        | √           | n/a (tx error)         |  √             | kernel panic | √            | √             |
| 4.18.0-25  (ubuntu 18)        | √           | n/a (tx error)         |  wrong dest ip | kernel panic | √            | √             |
| 4.18.0-227 (centos 8)         | √           | n/a (tx error)         |  wrong dest ip | kernel panic | √            | √             |
| 5.4.96-200 (centos 7 AltArch) | √           | n/a (tx error)         |  √             | kernel panic | √            | √             |
| 5.4.0-66   (ubuntu 18)        | √           | n/a (tx error)         |  √             | kernel panic | √            | √             |
| 5.11.1     (elrepo)           | √           | n/a (tx error)         |  √             | kernel panic | √            | √             |

### ipip encap by healthchecking_kern.o
| healthchecking_kern.o         | 4in4 | 4in6 | 6in4 | 6in6 |
| ----------------------------- | ---- |----- | ---- | ---- |
| 4.15.0-136 (ubuntu 18)        |  x   |  x   |  x   |  x   |
| 4.18.0-24  (ubuntu 18)        |  x   |  x   |  x   |  x   |
| 4.18.0-227 (centos 8)         |  √   |  √   |  √   |  √   |
| 5.4.96-200 (centos 7 AltArch) |  √   |  √   |  √   |  √   |
| 5.4.0-66   (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 5.11.1     (elrepo)           |  √   |  √   |  √   |  √   |

### gue encap by healthchecking_kern.o
| healthchecking_kern.o         | 4in4 | 4in6 | 6in4 | 6in6 |
| ----------------------------- | ---- |----- | ---- | ---- |
| 4.15.0-136 (ubuntu 18)        |  x   |  x   |  x   |  x   |
| 4.18.0-25  (ubuntu 18)        |  x   |  x   |  x   |  x   |
| 4.18.0-227 (centos 8)         |  √   |  √   |  √   |  √   |
| 5.4.96-200 (centos 7 AltArch) |  √   |  √   |  √   |  √   |
| 5.4.0-66   (ubuntu 18)        |  √   |  √   |  √   |  √   |
| 5.11.1     (elrepo)           |  √   |  √   |  √   |  √   |

### CODE REVIEW

File: build/fbcode_builder/manifests/libmnl [hunk 1]
```diff
@@ -9,7 +9,7 @@ libmnl-static
 libmnl-dev
 
 [download]
-url = http://www.lg.ps.pl/mirrors/ftp.netfilter.org/libmnl/libmnl-1.0.4.tar.bz2
+url = http://www.netfilter.org/pub/libmnl/libmnl-1.0.4.tar.bz2
```
Comment (udippant) [inline build/fbcode_builder/manifests/libmnl:12]: Can you submit separate PR for this?

File: katran/lib/bpf/pckt_encap.h [hunk 2]
```diff
@@ -218,19 +227,27 @@ static inline bool gue_encap_v4(struct xdp_md *xdp, struct ctl_value *cval,
...
+    pkt_bytes += sizeof(struct udphdr);
+  }
+
   create_udp_hdr(
     udph,
     sport,
     GUE_DPORT,
-    pkt_bytes + sizeof(struct udphdr),
+    pkt_bytes,
     GUE_CSUM);
```
Comment (udippant) [inline katran/lib/bpf/pckt_encap.h:243]: Can you rebase? We've changed the computation of csum for gue.

File: build/fbcode_builder/manifests/libmnl (hunk 1 above)
Comment (udippant) [inline build/fbcode_builder/manifests/libmnl:12]: Can you submit separate PR for this?

**Review by udippant — CHANGES_REQUESTED** [review]
You also need to update decap_* methods since the decap for ipv4 assume the inner packet to be v4 as well.

File: katran/lib/bpf/pckt_encap.h (hunk 2 above)
Comment (mchtech) [reply katran/lib/bpf/pckt_encap.h]: > You also need to update decap_* methods since the decap for ipv4 assume the inner packet to be v4 as well.

I have updated `(gue_)decap_(v4|v6)` methods in latest commit.



> Can you rebase? We've changed the computation of csum for gue.

updated.



In addition, there is a bug in `add_pseudo_ipv4_header` and `rem_pseudo_ipv4_header` in this commit (https://github.com/facebookincubator/katran/commit/b570f79a955e91a00bc80b208ce55892cc77e31c)

`iph->tot_len` contains the size of IPv4 header, but `udp length` field of pseudo header should be payload size (size  of (udp header + data) ).

It leads to udp checksum error when encap mode is IPv4-in-IPv6 (or IPv6-in-IPv4).

File: build/fbcode_builder/manifests/libmnl (hunk 1 above)
Comment (mchtech) [reply build/fbcode_builder/manifests/libmnl]: I have resubmited another PR #117

File: katran/lib/bpf/pckt_encap.h (hunk 2 above)
Comment (udippant) [reply katran/lib/bpf/pckt_encap.h]: Thanks for flagging. Yes, that looks like a bug indeed.

**Comment by udippant** [comment]
This PR is currently in internal tests and review. 

Meanwhile, I've pulled out our fix for the bug in the computation of GUE over ipv4 separately for easier test (and separate that from the rest of this change, which adds a new feature).

**Comment by mchtech** [comment]
rebase from master branch

**Comment by mchtech** [comment]
merge from master branch for `eth_hdr` struct changes: https://github.com/facebookincubator/katran/commit/74619b17d83f160aff60a9924139c1ba317a47d5

**Comment by avasylev** [comment]
Quite old PR, please create new one if you still want changes merged.
